// npm-failsafe/lib/Failsafe.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Failsafe = void 0;
const child_process_1 = require("child_process");
const readline = require("readline");
class Failsafe {
    constructor(logger, config, env) {
        this.logger = logger;
        this.config = config;
        this.env = env;
    }
    async run(scriptsName) {
        if (scriptsName.length === 0) {
            this.logger.error('failsafe', [
                `Please specify which npm scripts you'd like to run, for example:`,
                `  npm failsafe start test`
            ].join('\n'));
            return 1;
        }
        return scriptsName.reduce((previous, script_name) => {
            return previous
                .then(previous_exit_code => this.runScript(script_name)
                .then(current_exit_code => Math.max(previous_exit_code, current_exit_code)));
        }, Promise.resolve(0));
    }
    runScript(script_name) {
        return new Promise((resolve, reject) => {
            const npm = process.platform.startsWith('win32')
                ? `npm.cmd`
                : `npm`;
            const script = (0, child_process_1.spawn)(npm, [`run`, script_name], {
                cwd: this.config.cwd,
                env: {
                    'FORCE_COLOR': this.config.isTTY ? '1' : undefined,
                    ...this.env
                }
            });
            const stdout = readline.createInterface({ input: script.stdout }), stderr = readline.createInterface({ input: script.stderr });
            stdout
                .on('line', line => this.logger.info(script_name, line));
            stderr
                .on('line', line => this.logger.error(script_name, line));
            script.once('close', (code) => {
                stdout.close();
                stderr.close();
                this.logger.info('failsafe', `Script '${script_name}' exited with code ${code}`);
                resolve(code ?? 0);
            });
        });
    }
}
exports.Failsafe = Failsafe;
